### ISSUE TITLE
ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8010

### ISSUE THREAD

**Issue by txpet** [marklogic/java-client-api#855]
This is happening on java client api 4.0.3 and 9.0-3. Doesn't matter what I set context.mlserver to. I keep getting the reference to localhost in the stack trace. Please make this a high priority in that I cannot go to production until I can get this code working.

Below is the code:

```java
DatabaseClient client = DatabaseClientFactory.newClient(context.mlServer, context.mlStagePort,
				new DatabaseClientFactory.DigestAuthContext(context.mlUserName, context.mlPassword));

// get a flow manager
FlowManager flowManager = new FlowManager(dataHubConfig);

// retrieve the flow you wish to run
Flow harmonizeFlow = flowManager.getFlow(entityName, "harm_"+entityName, FlowType.HARMONIZE);


// build the flow runner
FlowRunner flowRunner = flowManager.newFlowRunner()
		.withFlow(harmonizeFlow)
		.withBatchSize(context.mlHarmonizeBatchSize)
		.withThreadCount(context.mlHarmonizeThreadCount)
		//.withOptions(options)
		.withSourceClient(client)
		.withDestinationDatabase(context.mlFinalDatabase)
		.onItemComplete((String jobId, String itemId) -> {
			System.out.println("Success: " + "harm_" + entityName + " JobId: " + jobId + " ItemId: " + itemId);
		}).onItemFailed((String jobId, String itemId) -> {
			System.err.println("Error: " + "harm_" + entityName + " JobId: " + jobId + " ItemId: " + itemId);
		});

// run the flow
JobTicket jobTicket = flowRunner.run();

// optionally wait for the flow to finish running
flowRunner.awaitCompletion();
System.out.println("end harm");
```

Below is the log file:

```
[INFO ]: dfprojectgit.harmonize_0_1.Harmonize - TalendJob: 'Harmonize' - Start.
[statistics] connecting to socket on port 10204
[statistics] connected
[INFO ]: com.marklogic.hub.HubConfig - Hub Project Dir: E:\\Users\\tpeterson\\MarkLogic\\Projects\\Transactions
[INFO ]: dfprojectgit.harmonize_0_1.Harmonize - tLoop_1 - Start to loop from 1 to 885 with a step of 1 (Increasing).
[INFO ]: org.apache.cxf.interceptor.LoggingOutInterceptor - Outbound Message
---------------------------
ID: 1
Address: http://phx-dihost-03.payflex.com:8010/LATEST/documents?uri=/SYSTEM/mljp/data_loader_boolean_type.xml
Http-Method: GET
Content-Type: 
Headers: {Accept=[application/xml]}
--------------------------------------
[INFO ]: org.apache.cxf.interceptor.LoggingInInterceptor - Inbound Message
----------------------------
ID: 1
Response-Code: 200
Encoding: UTF-8
Content-Type: application/xml; charset=utf-8
Headers: {connection=[Keep-Alive], Content-Length=[915], Content-type=[application/xml; charset=utf-8], Keep-Alive=[timeout=5], ML-Effective-Timestamp=[15108772803888044], Server=[MarkLogic], vnd.marklogic.document-format=[xml]}
Payload: <?xml version="1.0" encoding="UTF-8"?>
<envelope xmlns="http://marklogic.com/entity-services">
  <headers><source_system xmlns="">SYSTEM</source_system><ingest_timestamp xmlns="">2017-11-16T12:07:12.9270055-06:00</ingest_timestamp><ingest_source xmlns="">source_system</ingest_source><archived xmlns="">false</archived><deleted xmlns="">false</deleted></headers>
  <triples/>
  <instance>
    <info>
      <title>SYSTEM_MlJobProperties</title>
      <version>0.0.1</version>
    </info>
    <SYSTEM_MlJobProperties xmlns=""/>
  </instance>
  <attachments><controller xmlns="">
  <id>cbas_blueqa_data_loader_boolean_type</id>
  <source_system>CBAS</source_system>
  <environment>blueqa</environment>
  <table_name>data_loader_boolean_type</table_name>
  <entity_name>CBAS_DataLoaderBooleanType</entity_name>
  <last_harmonization>2017-11-16T12:09:58-0600</last_harmonization>
</controller></attachments>
</envelope>

--------------------------------------
[FATAL]: dfprojectgit.harmonize_0_1.Harmonize - tJavaRow_1 java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8010
com.marklogic.client.MarkLogicIOException: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8010
	at com.marklogic.client.impl.OkHttpServices.sendRequestOnce(OkHttpServices.java:697)
	at com.marklogic.client.impl.OkHttpServices.sendRequestOnce(OkHttpServices.java:690)
	at com.marklogic.client.impl.OkHttpServices.doGet(OkHttpServices.java:3965)
	at com.marklogic.client.impl.OkHttpServices.access$1300(OkHttpServices.java:165)
	at com.marklogic.client.impl.OkHttpServices$20.apply(OkHttpServices.java:3076)
	at com.marklogic.client.impl.OkHttpServices$20.apply(OkHttpServices.java:3074)
	at com.marklogic.client.impl.OkHttpServices.sendRequestWithRetry(OkHttpServices.java:719)
	at com.marklogic.client.impl.OkHttpServices.getIteratedResourceImpl(OkHttpServices.java:3079)
	at com.marklogic.client.impl.OkHttpServices.getIteratedResource(OkHttpServices.java:3058)
	at com.marklogic.client.impl.ResourceServicesImpl.get(ResourceServicesImpl.java:61)
	at com.marklogic.client.impl.ResourceServicesImpl.get(ResourceServicesImpl.java:57)
	at com.marklogic.hub.FlowManager.getFlow(FlowManager.java:232)
	at dfprojectgit.harmonize_0_1.Harmonize.tRESTClient_1Process(Harmonize.java:3861)
	at dfprojectgit.harmonize_0_1.Harmonize.tLoop_1Process(Harmonize.java:2391)
	at dfprojectgit.harmonize_0_1.Harmonize.runJobInTOS(Harmonize.java:4868)
	at dfprojectgit.harmonize_0_1.Harmonize.main(Harmonize.java:4507)
Caused by: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8010
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:240)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:158)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:256)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:134)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:113)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.burgstaller.okhttp.AuthenticationCacheInterceptor.intercept(AuthenticationCacheInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at com.marklogic.client.impl.OkHttpServices.sendRequestOnce(OkHttpServices.java:695)
	... 15 more
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:85)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:125)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:238)
	... 36 more
```

**Comment by sammefford** [comment]
@txpet, we'll look into this.  But if, as you say, this is impacting your ability to move into production, please work with support.  You can email support@marklogic.com if you haven't yet.  That's the best way to get the priority attention you need.

**Comment by sammefford** [comment]
Can you share information about how ML Server is configured?  We need to make sure you have a REST server on port 8010 configured with digest auth to match your code.  You're not trying to do SSL, correct? Also, can you confirm how many nodes you have in your cluster and that each one has the REST server configured?  Can you share the hostname you see in the middle top of the red gui on port 8001 for each host?  Can you confirm that each of those hostnames plus the hostname you're using for context.mlServer is accessible via ping from your client server?

Also, can you test doing a read with the DatabaseClient instance to see if it works outside the FlowRunner?  Since FlowRunner uses DMSDK it will try to connect to every host in your cluster, and the problem may arise with the configuration of one of those hosts.

In summary, it's very possible that a hostname isn't resolvable via DNS or you're just not configuring your ML server to match the way you've configured DatabaseClientFactory.

**Comment by vivekmuniyandi** [comment]
In addition to all these, can you please confirm how the 'dataHubConfig' object is got. I think there should be a properties file from which the data hub configuration parameters should be got. If there is no such file, gradle.properties should be the default. 

If you have set the mlHost in the properties file to localhost, this would happen. 

From the stack trace, I am seeing that, it is trying to get the flow from the server (which should be the datahub staging client) and this should have nothing to do with the sourceClient you are setting it to. 

Please check the properties file and correcting it should solve the problem in my opinion.

**Comment by txpet** [comment]
Sorry for the delay, I was on vacation and did not see the message until this morning. Changing the gradle.properties file worked. 

Thanks for your input.
